src/annotator/integrations/html-side-by-side.js

```diff
@@ -0,0 +1,171 @@
+  // since some pages, especially eg. in ebooks, may not have any paragraphs
+  // (eg. instead they may only contain tables or lists or headings).
+  const paragraphs = Array.from(root.querySelectorAll('p'))
+    .map(p => {
+      // Gather some data about them.
+      const rect = p.getBoundingClientRect();
+      const textLength = /** @type {string} */ (p.textContent).length;
+      return { rect, textLength };
+    })
+    .filter(({ rect }) => {
+      // Filter out hidden paragraphs
+      return rect.width > 0 && rect.height > 0;
+    })
+    // Select the paragraphs containing the most text.
+    .sort((a, b) => b.textLength - a.textLength)
+    .slice(0, 15);
+
+  // Let these paragraphs "vote" for what the left and right margins of the
+  // main content area in the document are.
+  paragraphs.forEach(({ rect }) => {
+    let leftVotes = leftMarginVotes.get(rect.left) ?? 0;
+    leftVotes += 1;
+    leftMarginVotes.set(rect.left, leftVotes);
+
+    let rightVotes = rightMarginVotes.get(rect.right) ?? 0;
+    rightVotes += 1;
+    rightMarginVotes.set(rect.right, rightVotes);
+  });
+
+  // Find the margin values with the most votes.
+  if (leftMarginVotes.size === 0 || rightMarginVotes.size === 0) {
+    return null;
+  }
+
+  const leftMargin = [...leftMarginVotes.entries()].sort((a, b) => b[1] - a[1]);
+  const rightMargin = [...rightMarginVotes.entries()].sort(
+    (a, b) => b[1] - a[1]
+  );
+
+  const [leftPos] = leftMargin[0];
+  const [rightPos] = rightMargin[0];
+
+  return { left: leftPos, right: rightPos };
+}
+
+/**
+ * @param {DOMRect} a
+ * @param {DOMRect} b
+ */
+function intersect(a, b) {
+  const left = Math.max(a.left, b.left);
+  const right = Math.min(a.right, b.right);
+  const top = Math.max(a.top, b.top);
+  const bottom = Math.min(a.bottom, b.bottom);
+  return new DOMRect(left, top, right - left, bottom - top);
+}
+
+/**
+ * Find content within an element to use as an anchor when scrolling.
+ *
+ * @param {Element} scrollRoot
+ * @return {Range|null} - Range to use as an anchor or `null` if a suitable
+ *   range could not be found
+ */
+function getScrollAnchor(scrollRoot) {
+  // Range representing the content whose position within the viewport we will
+  // try to maintain after running the callback.
+  let anchorRange = /** @type {Range|null} */ (null);
+
+  const viewport = intersect(
+    scrollRoot.getBoundingClientRect(),
+    new DOMRect(0, 0, window.innerWidth, window.innerHeight)
+  );
+  if (viewport.width < 0 || viewport.height < 0) {
+    // Element being scrolled is outside the viewport
+    return null;
+  }
+
+  // Create a range that includes the first word that is fully visible in the
+  // viewport.
+  //
+  // This currently visits every text node in `scrollRoot` and tests every
+  // possible anchor until it finds a suitable one. This could be optimized
+  // by skipping over elements that are entirely outside the viewport.
```

Comment:
It took me a bit to get my head around what I think this comment means. Maybe it's just the specific wording that's throwing me. 

I interpret what this loop does as: Walk text nodes in `scrollRoot` and find the first word whose bounding box is fully within the viewport. Return the `Range` representing that word as the scroll anchor.

Is that approximately correct? I have a sense you might say "But that's what the commenting already says." :) But for some reason I had to chew on it a bit.

Reply:
Regarding optimization: Can we talk about that a bit? If one were scrolled many screenfuls into a largish textual document the current approach...might take a while :)